crates/rust-analyzer/src/diagnostics/to_proto.rs

```diff
@@ -61,21 +61,39 @@ fn location(
     let file_name = resolve_path(config, workspace_root, &span.file_name);
     let uri = url_from_abs_path(&file_name);
 
-    // FIXME: this doesn't handle UTF16 offsets correctly
     let range = lsp_types::Range::new(
-        lsp_types::Position::new(
-            (span.line_start as u32).saturating_sub(1),
-            (span.column_start as u32).saturating_sub(1),
-        ),
-        lsp_types::Position::new(
-            (span.line_end as u32).saturating_sub(1),
-            (span.column_end as u32).saturating_sub(1),
-        ),
+        position(span, span.line_start, span.column_start),
+        position(span, span.line_end, span.column_end),
     );
 
     lsp_types::Location { uri, range }
 }
 
+fn position(
+    span: &DiagnosticSpan,
+    line_offset: usize,
+    column_offset: usize,
+) -> lsp_types::Position {
+    let line_index = line_offset - span.line_start;
+    let mut true_column_offset = column_offset;
+    if let Some(line) = span.text.get(line_index) {
+        let mut char_offset = 0;
+        for char in line.text.chars() {
```

Comment:
Character by character is slow for large inputs. Is it possible to get a fast path for ascii text?

Reply:
i made a simple benchmark, for all ascii text, this will save 60% 

```
if let Some(line) = span.text.get(line_index) {
  if text.chars().count() == text.len() {
          // all utf-8 
          return position{}
      }
 .....
}

```